### PR TITLE
feat!: add support for nullable and optional types in go 

### LINKED ISSUE #1418: Nullable types in golang
#### Reason/Context
Please try answering few of those questions
- Why we need this improvement?
Nullable strings are vital in many languages, also in go! In our current project we have put a lot of effort to support nullability between frontend and bff (graphql) - if it cannot be kept between services (asyncapi), it's a dealbreaker!
Currently `type: ["integer", "null"]` is translated to `interface{}` which is not typesafe. Other generators, e.g. https://github.com/99designs/gqlgen solves the problem by converting optional primitives to pointers, which are nullable. E.g. a type `string` becomes type `*string`.

https://github.com/asyncapi/modelina/commit/53edd32fcbc2dde8f9f36a0d344e2abc45b1fec2
This change addresses nullable types in typescript - could it be expanded to go?

#### Description
Please try answering few of those questions
- Will this be a breaking change?
Yes

### PR DESCRIPTION
**Pull request by Souvikns**
## Description

Adding support for nullable and required object for golang

## Related Issue

Fixes #1418

## Checklist
- [x] The code follows the project's coding standards and is properly linted (`npm run lint`).
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated to reflect the changes.
- [x] All tests pass successfully locally.(`npm run test`).

### CODE REVIEW

File: src/generators/go/GoConstrainer.ts [hunk 1]
```diff
@@ -48,6 +69,31 @@ export const GoDefaultTypeMapping: GoTypeMapping = {
...
+  type: string;
+}) {
+  let requried = false;
+  if (partOfProperty) {
+    if (partOfProperty.required) {
+      requried = partOfProperty.required;
+    } else {
+      requried = false;
+    }
+  }
```
Comment (jonaslagoni) [inline src/generators/go/GoConstrainer.ts:90]: Do you mind simplifying this if statement?  - and its called `required` not `requried` 🙂

File: test/generators/go/GoConstrainer.spec.ts [hunk 2]
```diff
@@ -184,3 +184,167 @@ describe('GoConstrainer', () => {
     });
   });
 });
+
+describe('nullable & requried properties', () => {
```
Comment (jonaslagoni) [inline test/generators/go/GoConstrainer.spec.ts:188]: For simplicity, I would recommend merging this grouping of tests with the regular tests, i.e. you have 4 tests under `describe('String', () => {` to keep it simple and a single place to look for the tests for specific type.

File: test/generators/go/GoGenerator.spec.ts [hunk 3]
```diff
@@ -168,7 +168,7 @@ describe('GoGenerator', () => {
         $id: 'Address',
         type: 'object',
         properties: {
-          street_name: { type: 'string' },
+          street_name: [{ type: 'string' }, { type: 'null' }],
```
Comment (jonaslagoni) [inline test/generators/go/GoGenerator.spec.ts:171]: This is not valid JSON Schema 🤔

**Review by jonaslagoni — CHANGES_REQUESTED** [review]

File: test/generators/go/GoConstrainer.spec.ts [hunk 4]
```diff
@@ -184,3 +184,167 @@ describe('GoConstrainer', () => {
...
+
+    test('requried', () => {
+      const model = new ConstrainedFloatModel('test', undefined, {}, '');
+      const type = GoDefaultTypeMapping.Float({
+        constrainedModel: model,
+        partOfProperty: {
+          required: true,
+          unconstrainedPropertyName: '',
+          property: model,
+          propertyName: ''
```
Comment (jonaslagoni) [inline test/generators/go/GoConstrainer.spec.ts:295]: This needs to be an instance of ConstrainedObjectPropertyModel 🙂

**Review by jonaslagoni — CHANGES_REQUESTED** [review]

**Review by jonaslagoni — APPROVED** [review]
Nice 👍

**Comment by Souvikns** [comment]
@jonaslagoni black box testing is failing, what to do for that.

File: src/generators/go/GoConstrainer.ts [hunk 5]
```diff
@@ -6,6 +6,7 @@ import { defaultModelNameConstraints } from './constrainer/ModelNameConstrainer'
 import { defaultPropertyKeyConstraints } from './constrainer/PropertyKeyConstrainer';
 import { defaultConstantConstraints } from './constrainer/ConstantConstrainer';
 import { GoTypeMapping } from './GoGenerator';
+import { ConstrainedMetaModel, ConstrainedObjectPropertyModel } from 'models';
```
Comment (jonaslagoni) [inline src/generators/go/GoConstrainer.ts:9]: This needs to change to relative path instead of absolute 🙂

**Review by jonaslagoni — CHANGES_REQUESTED** [review]

**Review by jonaslagoni — APPROVED** [review]
LGTM 👍

**Pull request merged**